specs/phase0/p2p-interface.md

```diff
@@ -1010,30 +1012,90 @@ Clients MAY connect to peers with the same `fork_digest` but a different `next_f
+    # nodes will transition from one shard to another throughout a period.
+    shuffling_bits = (node_id >> shuffling_bit_size) % (1 << NETWORK_SHARD_SHUFFLING_PREFIX_BITS)
+    # Calculates a multiplier that scales the shuffling prefix (assumed smaller
+    # than the rotation period) to be uniform throughout the entire rotation
+    # period. Can also be calculated as:
+    # EPOCHS_PER_SHARD_SUBSCRIPTION // 1 >> NETWORK_SHARD_SHUFFLING_PREFIX_BITS
+    shuffling_multiplier = EPOCHS_PER_SHARD_SUBSCRIPTION >> NETWORK_SHARD_SHUFFLING_PREFIX_BITS
+    # The epoch at which this node will rotate to a new shard
+    # This is distributed uniformly throughout the rotation period with a
+    # granularity based on the size of NETWORK_SHARD_SHUFFLING_PREFIX_BITS
+    epoch_transition = (
+        (shard_prefix + (shuffling_bits * shuffling_multiplier)) % EPOCHS_PER_SHARD_SUBSCRIPTION
+    )
+    # A seed which changes every rotation period (EPOCHS_PER_SHARD_SUBSCRIPTION)
+    # This enforces the rotation period and is staggered for each prefix so
+    # that nodes do not rotate from shards all at once
+    permutation_seed = hash(uint_to_bytes(uint64((epoch + epoch_transition) // EPOCHS_PER_SUBNET_SUBSCRIPTION)))
+    # The resulting value that ultimately defines the network shard.
     permutated_prefix = compute_shuffled_index(
-        node_id_prefix,
-        1 << ATTESTATION_SUBNET_PREFIX_BITS,
+        shard_prefix,
+        1 << NETWORK_SHARD_PREFIX_BITS,
         permutation_seed,
     )
-    return SubnetID((permutated_prefix + index) % ATTESTATION_SUBNET_COUNT)
+    return ShardID(permutated_prefix % NETWORK_SHARD_COUNT)
 ```
 
+The `compute_network_shard` function is designed with the following
+desirable properties:
+* It uses only the first set of bytes (defined by prefix_bytes_size) of the
+  node-id. This allows for efficient discovery searches, by allowing nodes to
+  search for specific nodes of network shards based on the kademilia XOR
+  metric.
+* Nodes will maintain a shard for EPOCHS_PER_SHARD_SUBSCRIPTION before
+  rotating to a new shard.
+* The rotation is staggered uniformly throughout the rotation period per shard.
+* No individual shard will suddenly rotate to another, rather a subset of nodes
+  per shard transition gradually throughout the rotation period.
+* The function is feasibly reversible. Prefixes can be calculated for desired
+  network shards for any given epoch, allowing nodes to search for these
+  prefixes.
+
+### Attestation Subnets
+
+The backbone structure for attestation subnets can be calculated from their
+network shard via the following:
+
 ```python
 def compute_subscribed_subnets(node_id: NodeID, epoch: Epoch) -> Sequence[SubnetID]:
-    return [compute_subscribed_subnet(node_id, epoch, index) for index in range(SUBNETS_PER_NODE)]
+    network_shard = compute_network_shard(node_id, epoch)
+    return [ShardId(network_shard + index) % ATTESTATION_SUBNET_COUNT for index in range(SUBNETS_PER_NODE)]
```

Comment:
If I'm understanding correctly, I think this is what you meant? (returning `SubnetID`s instead of `ShardID` here)

```suggestion
    return [SubnetID((network_shard + index) % ATTESTATION_SUBNET_COUNT) for index in range(SUBNETS_PER_NODE)]
```

Reply:
Yeah. Exactly. I guess assuming there is a conversion between the types.